src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java

```diff
@@ -170,31 +171,69 @@ public void close() {
+                boolean expressionUsesVariablesOfCurrentFlwor = !intersection.isEmpty();
+
+                if (expressionUsesVariablesOfCurrentFlwor) {
+                    throw new InvalidForClauseExpressionException(
+                            "An RDD-enabled for clause expression cannot reference the variables defined in the encapsulating, RDD-enabled FLWOR expression.",
+                            getMetadata().getExpressionMetadata()
+                    );
+                }
 
-            JavaRDD<Row> rowRDD = initialRdd.map(new ForClauseSerializeClosure());
+                // since no variable dependency to the current FLWOR expression exists for the expression
+                // evaluate the DF with the parent context and calculate the cartesian product
+                Dataset<Row> expressionDF;
+                if (_expression.isRDD()) {
+                    expressionDF = getDataFrameFromRDDExpression(context);
+                } else {
+                    expressionDF = getDataFrameFromLocalExpression(context);
+                }
 
-            // apply the schema to row RDD
-            return SparkSessionManager.getInstance().getOrCreateSession().createDataFrame(rowRDD, schema);
-        }
+                String inputDFTableName = "input";
+                String expressionDFTableName = "expression";
 
-        if (_child.isDataFrame()) {
-            Dataset<Row> df = this._child.getDataFrame(context, getProjection(parentProjection));
+                Dataset<Row> inputDF = this._child.getDataFrame(context, getProjection(parentProjection));
+                StructType inputSchema = inputDF.schema();
+                int duplicateVariableIndex = Arrays.asList(inputSchema.fieldNames()).indexOf(_variableName);
+                List<String> columnsToSelect = DataFrameUtils.getColumnNames(
+                    inputSchema,
+                    duplicateVariableIndex,
+                    null
+                );
 
-            StructType inputSchema = df.schema();
+                if (duplicateVariableIndex == -1) {
+                    columnsToSelect.add(_variableName);
+                } else {
+                    columnsToSelect.add(expressionDFTableName + "`.`" + _variableName);
+                }
+                String selectSQL = DataFrameUtils.getSQL(columnsToSelect, false);
+
+                inputDF.createOrReplaceTempView(inputDFTableName);
+                expressionDF.createOrReplaceTempView(expressionDFTableName);
+
+                return inputDF.sparkSession()
+                    .sql(
+                        String.format(
+                            "select %s from %s, %s",
+                            selectSQL,
+                            inputDFTableName,
+                            expressionDFTableName
+                        )
+                    );
+            }
 
+            Dataset<Row> df = this._child.getDataFrame(context, getProjection(parentProjection));
```

Comment:
Can we add a comment // the expression is locally evaluated

Reply:
Done